In [ ]:
!pip -q install langchain
!pip -q install openai==0.28.1
!pip -q install PyPDF2
!pip -q install tiktoken
!pip install cohere

In [ ]:
from google.colab import drive, files

drive.mount("/content/drive/", force_remount=False)

#Home directory:
%cd "/content/drive/My Drive/Tesi/"

## Analisi esplorativa dei dati

Carico tutti i file pdf in una lista , in pdf_list salvo solo il nome, mentre in pdf_readers salvo le pagine del pdf.

In [4]:
import os
import PyPDF2

def carica_pdf(cartella):

  pdf_list = []
  pdf_readers = []
  for file in os.listdir(cartella):
    if file.endswith(".pdf"):
      pdf_list.append(file)
      pdf_reader = PyPDF2.PdfReader(os.path.join(cartella, file))
      pdf_readers.append(pdf_reader.pages)

  return pdf_readers,pdf_list


if __name__ == "__main__":
  # Il percorso della cartella con i pdf.
  cartella = "./Documenti"

  # Carica tutti i pdf.
  pdf_readers,pdf_list = carica_pdf(cartella)

  # Stampa il numero di pdf caricati.
  print(len(pdf_readers))


26


In [5]:
pdf_list

['Scheda MicroPrestito della Regione Puglia - edizione 2021.pdf',
 "Scheda NIDI - Nuove iniziative d'impresa_ Strumento di ingegneria finanziaria.pdf",
 'Aiuti agli Investimenti delle Piccole e Medie Imprese (Titolo II Capo 3 Reg_ Regionale 17_2014)_br ___br __.pdf',
 'Aiuti Agli Investimenti delle PMI nel Settore Turistico - Alberghiero (Titolo II capo 6 Reg_ Reg_ 17_2014) _br ___br __.pdf',
 'Contratti di Programma per Grandi Imprese Singole o con PMI aderenti (Art_ 17 Reg_ Regionale 17_2014 e s_m_i_).pdf',
 'Programmi Integrati di Agevolazione  PIA TURISMO (Art_ 50 Reg_ Regionale 17_2014 e s_m_i_).pdf',
 'Programmi Integrati di Agevolazioni PIA Medie Imprese Singole o con Aderenti (Art_ 26 Reg_ Regionale 17_2014 e s_m_i_).pdf',
 'Programmi Integrati di Agevolazioni PIA Piccole Imprese (Art_ 27 Reg_ Regionale 17_2014 e s_m_i_).pdf',
 'Scheda Accreditamento Servizi per il Lavoro.pdf',
 'Scheda_ Piani Formativi Aziendali.pdf',
 'Scheda Efficientamento Energetico Edifici Pubblici.pdf',


Estraggo il testo da tutte le pagine del PDF.


In [6]:
documenti_dict = {}

for pdf, reader in zip(pdf_list, pdf_readers):
  testo = ""
  for i, page in enumerate(reader):
    testo += page.extract_text()
  documenti_dict[pdf] = testo

In [7]:
print(len(documenti_dict))

26


In [8]:
print(len(documenti_dict["Scheda MicroPrestito della Regione Puglia - edizione 2021.pdf"]))

6289


In [ ]:
print(documenti_dict['Scheda MicroPrestito della Regione Puglia - edizione 2021.pdf'])

  
 
14 ottobre  2021 Pag 1 di 2 Centro Servizi Sistema Puglia  
www.sistema.puglia.it  
 
 
Scheda: Avviso MicroPrestito  della  Regione  Puglia  
 
Informazioni  generali  Titolo  bando  MicroPrestito  della  Regione  Puglia  
Descrizione  L’Avviso  è stato  adottato  per far fronte  alle esigenze  delle  imprese  derivanti  dalle  
ripercussioni  economiche  causate  dall’emergenza  epidemiologia  Covid -19. La 
misura vuole garantire la continuità dell’operatività aziendale  e la salvaguardia 
dell’occupazione.  
Le disposizioni  dell’Avviso  sono  conformi  all’approvazione  dell’Aiuto  di Stato  SA. 
57021 -Italia -Covid19 regime quadro del 21 maggio 2020 da parte della  
Commissione  europea  ai sensi dell’art.  108 del TFUE e  s.m.i.  
Risultati  attesi  La misura  sostiene  l’accesso  al credito  attraverso  interventi  di micro -finanza  
che consistono  in mutui  concessi  a tasso zero  e senza  garanzie.  
Destinatari  Il sistema  economico  e produttivo  pugliese.  Chi può

In [ ]:
documenti_dict['Scheda MicroPrestito della Regione Puglia - edizione 2021.pdf']

'  \n \n14 ottobre  2021 Pag 1 di 2 Centro Servizi Sistema Puglia  \nwww.sistema.puglia.it  \n \n \nScheda: Avviso MicroPrestito  della  Regione  Puglia  \n \nInformazioni  generali  Titolo  bando  MicroPrestito  della  Regione  Puglia  \nDescrizione  L’Avviso  è stato  adottato  per far fronte  alle esigenze  delle  imprese  derivanti  dalle  \nripercussioni  economiche  causate  dall’emergenza  epidemiologia  Covid -19. La \nmisura vuole garantire la continuità dell’operatività aziendale  e la salvaguardia \ndell’occupazione.  \nLe disposizioni  dell’Avviso  sono  conformi  all’approvazione  dell’Aiuto  di Stato  SA. \n57021 -Italia -Covid19 regime quadro del 21 maggio 2020 da parte della  \nCommissione  europea  ai sensi dell’art.  108 del TFUE e  s.m.i.  \nRisultati  attesi  La misura  sostiene  l’accesso  al credito  attraverso  interventi  di micro -finanza  \nche consistono  in mutui  concessi  a tasso zero  e senza  garanzie.  \nDestinatari  Il sistema  economico  e produttivo 

Osservo il numero di caratteri per ogni documento

In [13]:
import tiktoken
def num_tokens_from_string(string: str, encoding_name: str) -> int:
    """Returns the number of tokens in a text string."""
    encoding = tiktoken.get_encoding(encoding_name)
    num_tokens = len(encoding.encode(string))
    return num_tokens


In [16]:
list_label = []
for chiave, valore in documenti_dict.items():
  list_label.append(chiave)
  num_tokens = num_tokens_from_string(valore, "r50k_base")

  print(f"La chiave '{chiave}' ha una lista con di {len(valore)} elementi. che corrispondono ad un numero di token {num_tokens}")

La chiave 'Scheda MicroPrestito della Regione Puglia - edizione 2021.pdf' ha una lista con di 6289 elementi. che corrispondono ad un numero di token 2644
La chiave 'Scheda NIDI - Nuove iniziative d'impresa_ Strumento di ingegneria finanziaria.pdf' ha una lista con di 16305 elementi. che corrispondono ad un numero di token 6771
La chiave 'Aiuti agli Investimenti delle Piccole e Medie Imprese (Titolo II Capo 3 Reg_ Regionale 17_2014)_br ___br __.pdf' ha una lista con di 25127 elementi. che corrispondono ad un numero di token 10847
La chiave 'Aiuti Agli Investimenti delle PMI nel Settore Turistico - Alberghiero (Titolo II capo 6 Reg_ Reg_ 17_2014) _br ___br __.pdf' ha una lista con di 26391 elementi. che corrispondono ad un numero di token 11486
La chiave 'Contratti di Programma per Grandi Imprese Singole o con PMI aderenti (Art_ 17 Reg_ Regionale 17_2014 e s_m_i_).pdf' ha una lista con di 22782 elementi. che corrispondono ad un numero di token 9093
La chiave 'Programmi Integrati di Agevo

In [ ]:
# # Retrieve the metadata of the PDF page
# metadata = loader.metadata

Splitto il testo dei documenti in chunk in modo da avere un numero massimo di token che posso passare in input al modello gpt

In [19]:
from langchain.text_splitter import RecursiveCharacterTextSplitter, CharacterTextSplitter


r_splitter = RecursiveCharacterTextSplitter(
    chunk_size=3796,
    chunk_overlap=0,
    separators=["\n\n", "\n", "(?<=\. )", " ", ""]
)

for doc in documenti_dict.keys():
  documenti_dict[doc] = r_splitter.split_text(documenti_dict[doc])
  for j in range(len(documenti_dict[doc])):
    documenti_dict[doc][j] = documenti_dict[doc][j].replace("\n","")
    documenti_dict[doc][j] = documenti_dict[doc][j].replace("✓","")


In [20]:
len(documenti_dict['Scheda MicroPrestito della Regione Puglia - edizione 2021.pdf'])

2

In [21]:
len1 = len(documenti_dict['Scheda MicroPrestito della Regione Puglia - edizione 2021.pdf'][0])
len3 = len('facendo riferimento al documento: Scheda MicroPrestito della Regione Puglia - edizione 2021.pdf')
documenti_dict['Scheda MicroPrestito della Regione Puglia - edizione 2021.pdf'][0]

'14 ottobre  2021 Pag 1 di 2 Centro Servizi Sistema Puglia  www.sistema.puglia.it    Scheda: Avviso MicroPrestito  della  Regione  Puglia   Informazioni  generali  Titolo  bando  MicroPrestito  della  Regione  Puglia  Descrizione  L’Avviso  è stato  adottato  per far fronte  alle esigenze  delle  imprese  derivanti  dalle  ripercussioni  economiche  causate  dall’emergenza  epidemiologia  Covid -19. La misura vuole garantire la continuità dell’operatività aziendale  e la salvaguardia dell’occupazione.  Le disposizioni  dell’Avviso  sono  conformi  all’approvazione  dell’Aiuto  di Stato  SA. 57021 -Italia -Covid19 regime quadro del 21 maggio 2020 da parte della  Commissione  europea  ai sensi dell’art.  108 del TFUE e  s.m.i.  Risultati  attesi  La misura  sostiene  l’accesso  al credito  attraverso  interventi  di micro -finanza  che consistono  in mutui  concessi  a tasso zero  e senza  garanzie.  Destinatari  Il sistema  economico  e produttivo  pugliese.  Chi può  partecipare  Benef

La possibile domanda a questo chunk è:

In [22]:
question = "Quali requisiti devono soddisfare le microimprese e i lavoratori autonomi in Puglia per partecipare al programma di MicroPrestito della Regione Puglia, mirato a mitigare le conseguenze economiche del Covid-19, e quali procedure devono seguire per presentare la domanda di partecipazione?"
len2 = len(question)

print(len1+len2+len3)

4056


4056 sono i caratteri mentre usando la funzione per sapere quanti token utilizzeremo è :

In [23]:
val = documenti_dict['Scheda MicroPrestito della Regione Puglia - edizione 2021.pdf'][0] + " " + question
num_tokens = num_tokens_from_string(val, "r50k_base")
print(num_tokens)

1539


La somma di queste 3 lunghezze sarà minore di 1600 token

#Ricerca di prompt e completion per il Fine-Tuning

Uso text-davinci-003 per analizzare tutti i chunk e creare un prompt che sia la domanda più il riferimento al documento di cui fa parte il chunk

In [ ]:
import openai
import json
import time

openai.api_key = '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'

data_to_save = []

for key, value in documenti_dict.items():
    if isinstance(value, list):
        for text in value:
            prompt_text = f"Genera una domanda basata sul seguente testo:\n{text}\nDomanda:"

            num_tokens_input = len(prompt_text) / 4
            print(f"Numero approssimativo di token dell'input: {num_tokens_input}")

            response = None
            while response is None:
                try:
                    response = openai.Completion.create(
                        engine="text-davinci-003",
                        prompt=prompt_text,
                        max_tokens=150,
                        n=1,
                        stop=None,
                        temperature=0.7
                    )
                except openai.error.RateLimitError as e:
                    print(f"Rate limit reached. Waiting for 20 seconds...")
                    time.sleep(20)

            question = response.choices[0].text.strip()
            print(question)

            data = {
                "prompt": question + f" in riferimento al documento: '{key}'",
                "completion": text
            }
            data_to_save.append(data)

            time.sleep(2)


with open('domande_e_testi.json', 'w') as json_file:
    json.dump(data_to_save, json_file, indent=4)


Inoltre qui usando sempre lo stesso modello cerco di migliorare il testo(la completion) ovvero la risposta tenendo in considerazione le domande generate precedentemente

In [ ]:
import json
with open('fine_tune_improved.json', 'r') as json_file:
    data_to_save = json.load(json_file)

In [ ]:
data_to_save = data_to_save[78:]

In [ ]:
import openai
import json
import time

openai.api_key = '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'

for data in data_to_save:
    prompt_text = f"Migliora il seguente testo:\n{data['completion']} rimanendo fedele alla domanda {data['prompt']}"

    response = None
    while response is None:
        try:
            response = openai.Completion.create(
                        engine="text-davinci-003",
                        prompt=prompt_text,
                        max_tokens=1024,
                        n=1,
                        stop=None,
                        temperature=0.7
            )
        except openai.error.RateLimitError as e:
            print(openai.error.RateLimitError)
            print(f"Rate limit reached. Waiting for 20 seconds...")
            time.sleep(2)

    improved_completion = response.choices[0].text.strip()
    print(improved_completion)
    # Estrai il numero di token utilizzati dalla risposta
    number_of_tokens_used = response['usage']['total_tokens']
    print(f"Numero di token utilizzati: {number_of_tokens_used}")
    data['completion'] = improved_completion

    time.sleep(2)


In riferimento al documento "Scheda Avviso Pubblico Diploma Professionale 2022.pdf", i requisiti necessari per essere ammessi alla presentazione delle proposte progettuali sono i seguenti: i Soggetti/Organismi formativi inseriti nell’Elenco Regionale degli Organismi accreditati secondo il vigente sistema di accreditamento per “Obbligo d’istruzione/diritto -dovere” alla data di presentazione della proposta progettuale, ex L.R. n. 15/2002 e ss. mm. e ii. devono aver assolto alle eventuali procedure di ricollocazione del personale, ai sensi delle vigenti norme contrattuali collettive, nonché rispettare quanto previsto dal Decreto Interministeriale del 29/11/2007 e devono essere in regola, ai sensi dell’art. 17 della Legge n.68/1999, in materia di disciplina del diritto al lavoro dei disabili.
Numero di token utilizzati: 1675
A partire da agosto 2019, il percorso formativo deve essere articolato in Unità di Competenza e/o Unità Formative capitalizzabili. La progettazione del quarto anno di

**Costo OpenAI per richiedere le domande : 3.14 dollari**,
**Costo OpenAI per migliorare il testo in base alle domande: 1.86 dollari**,
**Entrambi eseguiti con text-davinci-003**

Esporto il mio dataset in un file json.

In [ ]:
with open('prompt_and_completion_79_improved.json', 'w') as json_file:
    json.dump(data_to_save, json_file, indent=4)

In [ ]:
len_input = len(data['prompt']) + len("in riferimento al documento: 'Scheda MicroPrestito della Regione Puglia - edizione 2021.pdf'") + len(data['completion'])
num_tokens_input = len_input / 4
print(f"Numero approssimativo di token dell'input: {num_tokens_input}")


In [ ]:
len(data['completion'])

In [ ]:
data

In [ ]:
from transformers import BertTokenizerFast, BertForQuestionAnswering
from transformers import pipeline

tokenizer = BertTokenizerFast.from_pretrained("osiria/bert-italian-cased-question-answering")
model = BertForQuestionAnswering.from_pretrained("osiria/bert-italian-cased-question-answering")

pipeline_qa = pipeline("question-answering", model = model, tokenizer = tokenizer)
pipeline_qa(context = documenti_dict['Scheda MicroPrestito della Regione Puglia - edizione 2021.pdf'][1], question = questions[0])

# {'score': 0.9922313690185547, 'start': 28, 'end': 34, 'answer': 'Milano'}


In [ ]:
documenti_dict